protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SetACLCommandParser.java

```diff
@@ -43,10 +48,29 @@ public SetACLCommandParser(StatusResponseFactory statusResponseFactory) {
 
     @Override
     protected ImapMessage decode(ImapRequestLineReader request, Tag tag, ImapSession session) throws DecodingException {
-        final String mailboxName = request.mailbox();
-        final String identifier = request.astring();
-        final String rights = request.astring();
-        request.eol();
-        return new SetACLRequest(tag, mailboxName, identifier, rights);
+        try {
+            final SetACLRequest.MailboxName mailboxName = new SetACLRequest.MailboxName(request.mailbox());
+            final Username identifier = Username.of(request.astring());
+            final String editModeAndRights = request.astring();
+            request.eol();
+
+            MailboxACL.EditMode editMode = MailboxACL.EditMode.REPLACE;
+            if (StringUtils.isNotEmpty(editModeAndRights)) {
+                switch (editModeAndRights.charAt(0)) {
+                    case MailboxACL.ADD_RIGHTS_MARKER:
+                        editMode = MailboxACL.EditMode.ADD;
+                        break;
+                    case MailboxACL.REMOVE_RIGHTS_MARKER:
+                        editMode = MailboxACL.EditMode.REMOVE;
+                        break;
+                }
+            }
```

Comment:
Method extraction to avoid reallocation please

Reply:
Can we leverage switch expressions/yield keyword to avoid reallocation here?

Reply:
Yes of course!

Though method extraction might still make sens to avoid too long of a method here...